### PR TITLE
UBI Ruby builder image

### LINKED ISSUE #35: UBI Ruby builder image
Add a new ruby builder image based on UBI image in order to support our conjur UBI flavour.

### PR DESCRIPTION
**Pull request by guygiat**
### Summary
It seems that openssl version that comes out of the box from the ubi base os repo is already FIPS compliance:
https://www.redhat.com/en/blog/how-rhel-8-designed-fips-140-2-requirements
We need to figure out if the Postgres client that we install using dnf/yum is FIPS as well as it will use the openssl within the os.

### What does this PR do?
- Add support to build ruby from source on UBI base image

### What ticket does this PR close?
Resolves #35 

### Checklists
- [x] Add new tests for ubi

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] This PR does not require updating any documentation, or
- [x] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs

### CODE REVIEW

**Comment by izgeri** [comment]
To consume this in cyberark/conjur#1895, you will need to publish it as well, right? That is, you'll need to add a `push.sh` script to this directory and publish it to DockerHub / the internal registry as you can see is being done for other images here: https://github.com/cyberark/conjur-base-image/blob/master/Jenkinsfile#L61-L80

I see the existing images publish to [`cyberark/ubuntu-ruby-fips`](https://hub.docker.com/r/cyberark/ubuntu-ruby-fips) and [`cyberark/phusion-ruby-fips`](https://hub.docker.com/r/cyberark/phusion-ruby-fips) - which I suppose implies that this image should be published to a new DockerHub page `cyberark/ubi-ruby` or `cyberark/ubi-ruby-fips` (?)

I'd be glad for your thoughts about next steps here - thanks for sticking with us as we figure out the best path forward to get this image built.

**Comment by guygiat** [comment]
You are correct, as of now the ubi is not yet fips compliance so I think that until we'll make it fips we can push it to `cyberark/ubi-ruby` WDYT?

**Comment by guygiat** [comment]
You are correct, as of now the ubi is not yet fips compliance so I think that until we'll make it fips we can push it to `cyberark/ubi-ruby` WDYT?

**Comment by izgeri** [comment]
Sure. Is this a valid short / long description? Can you add any more details?
> Combines a base Ubuntu image with Ruby

>  ubi-ruby combines a base UBI image with Ruby [compiled against the FIPS 140-2 compliant OpenSSL module?]. This image includes the following packages:
>
>    [OpenSSL version: 1.0.2u build with FIPS 140-2 compliant OpenSSL module version: 2.0.16?]
>    Ruby version: 2.5 [compiled against the FIPS 140-2 compliant OpenSSL module?]
>    Postgres client version: 12-12.2 [compiled against the FIPS 140-2 compliant OpenSSL module?]
>    Bundler version: 2.1.4
>
> Source code: https://github.com/cyberark/conjur-base-image

I put things in brackets that I wasn't sure should stay - can you please verify the versions are correct too?

File: ubi-ruby-builder/Dockerfile
Comment (sgnn7) [inline ubi-ruby-builder/Dockerfile:11]: This line should be on a separate `RUN` since you're downloading something very differently from the rest of the packages.

**Review by sgnn7 — COMMENTED** [review]
@guygiat Left a small comment but it seems ok though all of @izgeri comments are still valid.

File: ubi-ruby-builder/Dockerfile
Comment (izgeri) [inline ubi-ruby-builder/Dockerfile:6]: Why is this different from the other builder Dockerfiles? In particular, it's missing
```
# OpenSSL Builder
FROM openssl-builder:$OPENSSL_BUILDER_TAG as OpenSSL-builder
```

and

```
ENV LD_LIBRARY_PATH="/usr/local/ssl/lib"

COPY --from=OpenSSL-builder /usr/local/ssl/ /usr/local/ssl/

RUN sed -Ei 's/^# deb-src /deb-src /' /etc/apt/sources.list
```

Additionally, should the `UBI_VERSION` be an arg, the way that phusion and ubuntu versions are args in the other images?

File: ubi-ruby-builder/build.sh
Comment (izgeri) [inline ubi-ruby-builder/build.sh:10]: if we add UBI_VERSION and OPENSSL_BUILDER_TAG args, they'd need to be added here too

File: Jenkinsfile
Comment (izgeri) [inline Jenkinsfile:111]: did you try to use the ubuntu or phusion test scripts out of the box and see how it looked for this image?

File: ubi-ruby-fips/Description.md
Comment (izgeri) [inline ubi-ruby-fips/Description.md:6]:
```suggestion
* Ruby version `2.5`: compiled against the FIPS 140-2 compliant OpenSSL module.
```
I think we could leave this at 2.5 like the other images do

File: ubi-ruby-fips/Description.md
Comment (izgeri) [inline ubi-ruby-fips/Description.md:8]: do you not want to mention the OpenSSL and PG versions as the other images do?
```
* OpenSSL version `1.0.2u`: built with  FIPS 140-2 compliant OpenSSL module version `2.0.16`.
* Postgres client version `10-10.14`: compiled against the FIPS 140-2 compliant OpenSSL module.
```

File: ubi-ruby-fips/Dockerfile
Comment (izgeri) [inline ubi-ruby-fips/Dockerfile:6]: this is much more lightweight than the other image Dockerfiles; is there a reason not to copy files
```
FROM openssl-builder:$OPENSSL_BUILDER_TAG as OpenSSL-builder
FROM postgres-client-builder:$PG_BUILDER_TAG as postgres-client-builder
```

the same way the other images do?

**Review by izgeri — CHANGES_REQUESTED** [review]
I left a bunch of comments that I'd like to at least get more info about why you left out the OpenSSL / PG steps in your image builds.

Mainly I'd like to understand if you have strong reasoning for not including them, or if it's a timing / testing issue and a small effort from R&D could help to add these in to make this image more comparable to the other images published by this project.

File: Jenkinsfile
Comment (guygiat) [reply Jenkinsfile:66]: The original `buildTestAndScanImage` have test.sh that can't run on the ubi image because it searches for specific outputs for openssl commands and ubuntu specific commands, that is the reason I created a dedicated method that with a comment out of the test phase till we'll have a specific test for ubi.

File: ubi-ruby-builder/Dockerfile
Comment (guygiat) [reply ubi-ruby-builder/Dockerfile:6]: We can't use the openssl-builder because it compiled on top of ubuntu base image.
The UBI 8 image is already baked with FIPS openssl so we don't need to compile a dedicated openssl from ubi for that.
About `UBI_VERSION` it's not the same as ubuntu versioning, the ubi8/ubi repo will not change as the ubi 8 versions will be upgraded afaik.

File: Jenkinsfile
Comment (guygiat) [reply Jenkinsfile:111]: The test scripts searches for specific outputs for openssl commands and ubuntu (apt) specific commands so it will break.

File: ubi-ruby-fips/Description.md
Comment (guygiat) [reply ubi-ruby-fips/Description.md:8]: The UBI doesn't have a specific version of openssl and Postgres client because we don't compile a dedicated version of it for UBI.
Because openssl on UBI 8 is FIPS out of the box it will be updated with the latest version they are pushing the the ubi 8 base os yum repo.

File: ubi-ruby-fips/Dockerfile
Comment (guygiat) [reply ubi-ruby-fips/Dockerfile:6]: We don't need to build it separately because it comes FIPS on UBI 8 out of the box.

**Review by aloncarmel111 — COMMENTED** [review]
@guygiat  Are we want to ensure UBI force FIPS140-2 requirements? if so we need enable FIPS mode by running the following command  `fips-mode-setup --enable`

**Comment by guygiat** [comment]
@aloncarmel111 I thought about enabling FIPS on UBI but from what I saw [here](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/security_hardening/using-the-system-wide-cryptographic-policies_security-hardening#enabling-fips-mode-in-a-container_using-the-system-wide-cryptographic-policies) in order to do that on the container level the host itself must be configured in FIPS mode.

File: Jenkinsfile
Comment (guygiat) [reply Jenkinsfile]: In my last commit I added a test and reverted back to original buildTestAndScanImage :)

File: ubi-ruby-builder/Dockerfile
Comment (guygiat) [reply ubi-ruby-builder/Dockerfile:11]: Added `UBI_VERSION` argument it my last commit :)

File: test-ubi.yml
Comment (izgeri) [inline test-ubi.yml:1]: Just noting that these tests are different from test.yml - @guygiat it might be good to note here why UBI doesn't test these cases
```
<   - name: "OpenSSL denies non FIPS compliant algorithms"
<     command: "openssl"
<     args:
<       - md5
<       - /etc/passwd
<     exitCode: 1
<     expectedError: [".*disabled for fips.*"]
<   # holded packages
<   - name: "packages are on hold"
<     command: "apt-mark"
<     args:
<       - "showhold"
<     expectedOutput:
<       - ".*ca-certificates.*"
<       - ".*libssl1.0.0.*"
<       - ".*libssl1.1.*"
<       - ".*openssl.*"
<   - name: "Ruby denies FIPS compliant algorithms"
<     command: "ruby"
<     args:
<       - -ropenssl
<       - -e
<       - 'OpenSSL.fips_mode = true'
<       - -ea
<       - 'puts(OpenSSL::Digest::MD5.new)'
<     exitCode: 1
<     expectedError: [".*disabled for fips.*"]
```

**Review by izgeri — APPROVED** [review]
Left a comment that we should note why we're missing the test cases, but other than that this LGTM

@guygiat I'm working on getting the dockerhub repo for this, I'll hold off on merging until that is up. Thanks!!

File: test-ubi.yml
Comment (guygiat) [reply test-ubi.yml:1]: - The openssl md5 on UBI is supported because FIPS needs be enforced on the host level in order to disable md5 on openssl.
-  On the ubuntu base image we hold the openssl, libssl, ca-certificates packages so we can test it's actually being hold. For UBI I didn't do it because we install those packages from the ubi repo.

Comment (aloncarmel111) [reply test-ubi.yml:1]: Please add the OPENSSL_FIPS=1 environment variable, this will enable FIPS mode for an openssl command.

Comment (izgeri) [reply test-ubi.yml:1]: @guygiat I'm still waiting on the dockerhub project - can you make this change that @aloncarmel111 requested before we merge? ^

Comment (guygiat) [reply test-ubi.yml:1]: @aloncarmel111 When I set the OPENSSL_FIPS=1 in ubi I'm getting `FIPS mode not supported.` when trying to run openssl commands, I think this environment variable is only supported on ubuntu/other flavours than redhat.
Are you familiar with this?

Comment (aloncarmel111) [reply test-ubi.yml:1]: @guygiat Seems that to enable container FIPS mode the host should also set to in FIPS mode see [here](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/security_guide/chap-federal_standards_and_regulations#sec-Enabling-FIPS-Mode).
The ruby flag `fips_mode` should works and ruby should deny MD5 algorithm can you add this verification please.

Comment (guygiat) [reply test-ubi.yml:1]: Sure no problem.
@izgeri To summarise: I've added the ruby md5 check so we'll drop only the openssl md5 test and the package hold test.

File: CHANGELOG.md
Comment (izgeri) [inline CHANGELOG.md:15]:
```suggestion
- UBI-based image to support Conjur server running on OpenShift. Uses a builder image to install Ruby
  compiled with FIPS-enabled OpenSSL. 
  [cyberark/conjur-base-image#34](https://github.com/cyberark/conjur-base-image/pull/34)
```

**Review by izgeri — APPROVED** [review]

**Pull request merged**
